packages/react/src/floating-ui-react/hooks/useDismiss.ts

```diff
@@ -482,227 +522,286 @@ export function useDismiss(
         }),
       }),
-    }),
-    [closeOnEscapeKeyDown, store, referencePress, referencePressEvent],
-  );
+    };
+  }
 
-  const handlePressedInside = useStableCallback((event: React.MouseEvent) => {
-    const target = getTarget(event.nativeEvent) as Element | null;
-    if (!contains(store.select('floatingElement'), target) || event.button !== 0) {
-      return;
+  public getFloatingProps(): HTMLProps {
+    if (!this.settings.enabled) {
+      return EMPTY_OBJECT;
     }
-    endedOrStartedInsideRef.current = true;
-  });
 
-  const floating: ElementProps['floating'] = React.useMemo(
-    () => ({
-      onKeyDown: closeOnEscapeKeyDown,
-      onMouseDown: handlePressedInside,
-      onMouseUp: handlePressedInside,
-      onPointerDownCapture: markInsideReactTree,
-      onMouseDownCapture: markInsideReactTree,
-      onClickCapture: markInsideReactTree,
-      onMouseUpCapture: markInsideReactTree,
-      onTouchEndCapture: markInsideReactTree,
-      onTouchMoveCapture: markInsideReactTree,
-    }),
-    [closeOnEscapeKeyDown, handlePressedInside, markInsideReactTree],
+    return {
+      onKeyDown: this.closeOnEscapeKeyDown,
+      onMouseDown: this.handlePressedInside,
+      onMouseUp: this.handlePressedInside,
+      onPointerDownCapture: this.markInsideReactTree,
+      onMouseDownCapture: this.markInsideReactTree,
+      onClickCapture: this.markInsideReactTree,
+      onMouseUpCapture: this.markInsideReactTree,
+      onTouchEndCapture: this.markInsideReactTree,
+      onTouchMoveCapture: this.markInsideReactTree,
+    };
+  }
+
+  public dispose() {
+    this.cancelDismissOnEndTimeout.clear();
+    this.clearInsideReactTreeTimeout.clear();
+  }
+}
+
+/**
+ * Closes the floating element when a dismissal is requested — by default, when
+ * the user presses the `escape` key or outside of the floating element.
+ * @see https://floating-ui.com/docs/useDismiss
+ */
+export function useDismiss(
+  context: FloatingRootContext | FloatingContext,
+  props: UseDismissProps = {},
+): ElementProps {
+  const store = 'rootStore' in context ? context.rootStore : context;
+  const tree = useFloatingTree(props.externalTree);
+
+  const controllerParameters = {
+    ...props,
+    tree,
+  };
+
+  const controller = useRefWithInit(
+    () => new DismissInteractionController(store, controllerParameters),
+  ).current;
+
+  const [referenceProps, setReferenceProps] = React.useState<HTMLProps>(
+    controller.getReferenceProps(),
   );
+  const [floatingProps, setFloatingProps] = React.useState<HTMLProps>(
+    controller.getFloatingProps(),
+  );
+
+  // Update controller settings on each render
+  useIsoLayoutEffect(() => {
+    controller.updateSettings(controllerParameters);
+  }, Object.values(controllerParameters));
+
+  controller.onPropsChange = () => {
+    setReferenceProps(controller.getReferenceProps());
+    setFloatingProps(controller.getFloatingProps());
+  };
+
+  controller.useSetup();
+
+  React.useEffect(() => {
```

Comment:
Nitpick, but here you can save one function allocation:


```suggestion
  React.useEffect(controller.disposeEffect);
```

Where `disposeEffect` is:

```ts
  public disposeEffect() {
    return () => {
      this.cancelDismissOnEndTimeout.clear();
      this.clearInsideReactTreeTimeout.clear();
    }
  }
```